homeassistant/components/remote/xiaomi_miio.py

```diff
@@ -0,0 +1,261 @@
+        self._state = False
+        self._commands = commands
+
+    @property
+    def name(self):
+        """Return the name of the remote."""
+        return self._name
+
+    @property
+    def device(self):
+        """Return the remote object."""
+        return self._device
+
+    @property
+    def hidden(self):
+        """Return if we should hide entity."""
+        return self._is_hidden
+
+    @property
+    def slot(self):
+        """Return the slot to save learned command."""
+        return self._slot
+
+    @property
+    def timeout(self):
+        """Return the timeout for learning command."""
+        return self._timeout
+
+    @property
+    def is_on(self):
+        """Return False if device is unreachable, else True."""
+        from miio import DeviceException
+        try:
+            self.device.info()
+            return True
+        except DeviceException:
+            return False
+
+    @property
+    def should_poll(self):
+        """We should not be polled for device up state."""
+        return False
+
+    @property
+    def device_state_attributes(self):
+        """Hide remote by default."""
+        if self._is_hidden:
+            return {'hidden': 'true'}
+        else:
+            return
+
+    # pylint: disable=R0201
+    @asyncio.coroutine
+    def async_turn_on(self, **kwargs):
+        """Turn the device on."""
+        _LOGGER.error("Device does not support turn_on, " +
+                      "please use 'remote.send_command' to send commands.")
+
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs):
+        """Turn the device off."""
+        _LOGGER.error("Device does not support turn_off, " +
+                      "please use 'remote.send_command' to send commands.")
+
+    # pylint: enable=R0201
+    @asyncio.coroutine
+    def _send_command(self, payload):
+        """Send a command."""
+        from miio import DeviceException
+
+        _LOGGER.debug("Sending payload: '%s'", payload)
+        try:
+            yield from self.hass.async_add_job(
+                self.device.play, payload, None)
+            return True
+        except DeviceException as ex:
+            _LOGGER.error(
+                "Transmit of IR command failed, %s, exception: %s",
+                payload, ex)
+            return False
+
+    @asyncio.coroutine
+    def async_send_command(self, command, **kwargs):
+        """Wrapper for _send_command."""
+        num_repeats = kwargs.get(ATTR_NUM_REPEATS)
+
+        delay = kwargs.get(ATTR_DELAY_SECS, DEFAULT_DELAY_SECS)
+
+        for _ in range(num_repeats):
+            for payload in command:
+                if payload in self._commands:
+                    for local_payload in self._commands[payload][CONF_COMMAND]:
+                        yield from self._send_command(local_payload)
+                else:
+                    yield from self._send_command(payload)
+                yield from asyncio.sleep(delay, loop=self.hass.loop)
```

Comment:
Can we offload this repeat stuff into library or do we need this repeat code? If the device is not response, we could try it later.

Reply:
Repeat is not about devices not responding, this code snippet from remote.harmony was linked by @syssi earlier: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/remote/harmony.py#L209-L222

num_repeats and delay are optional parameters for remote.send_command